File: python/olive.py
from map import * ## "Parallel Map Algebra" package

f = read('flies')   # outbreak of olive fruit flies
o = read('olives')  # olive fields 
t = read('temper')  # avg. annual temperature
r = read('rain')    # avg. annual rain
w = read('rivers')  # rivers network
N = 50              # number of time steps

tT = [0,0.1,0.2,0.3,0.4,0.5,0.6,0.7,0.8,0.9,1.0] # temp table
rT = [0,.05,0.1,.15,0.2,.25,0.3,.35,0.4,.45,0.5] # rain table

def spread(f) :
	lst = [ 2, 3, 3, 3, 2,  #                   ▒ █ █ █ ▒
		    3, 2, 2, 2, 3,  # spreading         █ ▒ ▒ ▒ █
		    3, 2, 1, 2, 3,  # density      -->  █ ▒ · ▒ █
		    3, 2, 2, 2, 3,  # distribution      █ ▒ ▒ ▒ █
		    2, 3, 3, 3, 2 ] #                   ▒ █ █ █ ▒
	msk = Mask([5,5],lst) # 5x5 mask of weights
	return convolve(f,msk) / 61 # normalization

for i in range(N) :
	s = 0.5*w + pick(r,rT)  # suitability of cells
	f = spread(f)           # migration of flies
	b = o * s * (1-f)*f*4   # birth rate
	f = f + rand() * 0.1*b  # stochastic reproduction
	d = pick(t,tT) * f      # death rate
	f = f - rand() * 0.1*d  # stochastic extinction

write(f,'output')

File: bin/olive.cpp
/**
 * @file	urban.cpp
 * @author	Jesús Carabaño Bravo <jcaraban@abo.fi>
 *
 * Cellular automata for olive fly
 */

#include "../map.hpp"
using namespace map;


int main(int argc, char **argv)
{	
	//// Arguments

	assert(argc > 5);

	//// Configuration

	//setupDevices("",DEV_GPU,"");
	setupDevices("Intel",DEV_CPU,"");

	//// Inputs

	auto f = read(argv[1]); // number of flies
	auto o = read(argv[2]); // olive grove
	auto t = read(argv[3]); // avg. annual temperature
	auto r = read(argv[4]); // avg. annual rain
	auto w = read(argv[5]); // rivers network
	Raster s, br, dr;

	auto N  = (argc > 7) ? atoi(argv[7]) : 1 ;

	std::vector<VariantType> tT = {0.0f,0.1f,0.2f,0.3f,0.4f,0.5f,0.6f,0.7f,0.8f,0.9f,1.0f,1.0f};
	std::vector<VariantType> rT = {0.5f,0.45f,0.4f,0.35f,0.3f,0.25f,0.2f,0.15f,0.1f,0.05f,0.0f};

	auto spread = [](Raster f) {
		Array<int> h = { 2, 3, 3, 3, 2,
						 3, 2, 2, 2, 3,
						 3, 2, 1, 2, 3,
						 3, 2, 2, 2, 3,
						 2, 3, 3, 3, 2 };
		Mask H = Mask({5,5}, h);
		return convolve(f,H) / 61.0f;
	};

	// Main Loop

	for (int j=0; j<N; j++) {
		s = 0.5f*w + pick(r,rT);
		f = spread(f);
		br = o * s * (1-f)*f*4;
		f = f + 0.1f*br * rand(f+j);
		dr = pick(t,tT) * f;
		f = f - 0.1f*dr * rand(f+j);
	}

	f = stats(f);
	write(f,argv[6]);
}
